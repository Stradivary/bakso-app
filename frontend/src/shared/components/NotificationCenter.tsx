import {
  Button,
  Card,
  Indicator,
  Popover,
  ScrollArea,
  Stack,
  Text,
} from "@mantine/core";
import { useDisclosure } from "@mantine/hooks";
import dayjs from "dayjs";
import { Bell } from "lucide-react";
import { useCallback } from "react";
import { useAuth } from "../contexts/useAuth";
import { useNotifications } from "../hooks/useNotification";
import { NearbyUser } from "../models/user.types";

export function NotificationCenter({ users }: { users: NearbyUser[]; }) {
  const { user } = useAuth();
  const { notifications, unreadCount, markAsRead } = useNotifications(user?.id);
  const [opened, { toggle }] = useDisclosure();

  const handleNotificationClick = useCallback(async (notificationId: string) => {
    await markAsRead(notificationId);
  }, [markAsRead]);

  return (
    <Popover
      opened={opened}
      position="bottom-end"
      shadow="md"
    >
      <Popover.Target>
        <Indicator disabled={unreadCount === 0} processing
          size={12} >
          <Button
            variant="white"
            size="compact-lg"
            onClick={() => toggle()}
          >
            <Bell size={24} />
          </Button>
        </Indicator>
      </Popover.Target>

      <Popover.Dropdown w={250} style={{ zIndex: 800 }}>
        <Stack gap="xs">
          <Text fw={500}>Notifikasi</Text>
          <ScrollArea style={{ height: 300 }}>
            <Stack gap="xs">
              {notifications.length === 0 ? (
                <Text c="dimmed" ta="center">
                  Tidak ada notifikasi
                </Text>
              ) : (
                notifications.map((notification) => {
                  const buyer = users?.find((u) => u.id === notification.buyer_id);
                  return (
                    <Card
                      key={notification.id}
                      p="xs"
                      withBorder
                      onClick={() => handleNotificationClick(notification.id)}
                      style={{
                        backgroundColor: notification.is_read ? undefined : '#f0f9ff',
                        cursor: 'pointer'
                      }}
                    >
                      <Text size="sm">
                        Seorang pelanggan mencolekmu: {buyer?.name ?? 'Unknown User'}
                      </Text>
                      <Text size="xs" c="dimmed">
                        {notification.expiry_at
                          ? dayjs(notification.expiry_at).format("DD MMMM YYYY\n HH:mm")
                          : 'No date'}
                      </Text>
                    </Card>
                  );
                })
              )}
            </Stack>
          </ScrollArea>
        </Stack>
      </Popover.Dropdown>
    </Popover>
  );
}