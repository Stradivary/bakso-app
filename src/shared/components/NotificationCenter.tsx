import {
  ActionIcon,
  Button,
  Card,
  Indicator,
  Popover,
  ScrollArea,
  Stack,
  Text,
  Group,
} from "@mantine/core";
import { useDisclosure } from "@mantine/hooks";
import dayjs from "dayjs";
import { Bell, CheckCheck } from "lucide-react";
import { useCallback } from "react";
import { Notification } from "../hooks/useNotification";

export function NotificationCenter({
  notifications,
}: Readonly<{ notifications?: Notification[] }>) {
  const [opened, { toggle }] = useDisclosure();

  const handleNotificationClick = useCallback(
    async (notificationId: string) => {
      const notification = notifications?.find(
        (note) => note.id === notificationId,
      );
      if (!notification) return;
      notification.is_read = true;
    },
    [notifications],
  );

  const handleClearNotifications = useCallback(() => {
    notifications?.forEach((note) => (note.is_read = true));
  }, [notifications]);

  return (
    <Popover opened={opened} position="bottom-end" shadow="md">
      <Popover.Target>
        <Indicator
          disabled={notifications?.filter((note) => !note.is_read).length === 0}
          processing
          size={12}
        >
          <Button
            data-testid="notification-center"
            variant="white"
            size="compact-lg"
            onClick={() => toggle()}
          >
            <Bell size={24} />
          </Button>
        </Indicator>
      </Popover.Target>

      <Popover.Dropdown w={250} style={{ zIndex: 800 }}>
        <Stack gap="xs">
          <Group justify="space-between" align="center">
            <Text fw={500}>Notifikasi</Text>
            <ActionIcon
              variant="transparent"
              size="xs"
              onClick={() => handleClearNotifications()}
            >
              <CheckCheck size={20} />
            </ActionIcon>
          </Group>
          <ScrollArea style={{ height: 300 }}>
            <Stack gap="xs">
              {notifications?.length === 0 ? (
                <Text c="dimmed" ta="center">
                  Tidak ada notifikasi
                </Text>
              ) : (
                notifications?.map((notification) => {
                  return (
                    <Card
                      key={notification.id}
                      p="xs"
                      withBorder
                      onClick={() => handleNotificationClick(notification.id)}
                      style={{
                        backgroundColor: notification.is_read
                          ? undefined
                          : "#f0f9ff",
                        cursor: "pointer",
                      }}
                    >
                      <Text size="sm">
                        Seorang pelanggan mencolekmu:{" "}
                        {notification.buyer_name ?? "Unknown User"}
                      </Text>
                      <Text size="xs" c="dimmed">
                        {notification.expiry_at
                          ? dayjs(notification.expiry_at).format(
                              "DD MMMM YYYY\n HH:mm",
                            )
                          : "No date"}
                      </Text>
                    </Card>
                  );
                })
              )}
            </Stack>
          </ScrollArea>
        </Stack>
      </Popover.Dropdown>
    </Popover>
  );
}
