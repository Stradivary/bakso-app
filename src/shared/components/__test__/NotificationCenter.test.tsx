import { MantineProvider } from "@mantine/core";
import { render } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { NotificationCenter } from "../NotificationCenter";

describe("NotificationCenter", () => {
  const mockNotifications = [
    {
      id: "1",
      buyer_name: "John Doe",
      is_read: false,
      expiry_at: "2024-01-01T00:00:00Z",
      buyer_id: " ",
      created_at: " ",
      seller_id: " ",
    },
    {
      id: "2",
      buyer_name: "Jane Doe",
      is_read: true,
      expiry_at: "2024-01-02T00:00:00Z",
      buyer_id: " ",
      created_at: " ",
      seller_id: " ",
    },
  ];

  it("renders notification center", () => {
    const { container } = render(
      <MantineProvider>
        <NotificationCenter notifications={mockNotifications} />
      </MantineProvider>,
    );
    expect(container).toMatchSnapshot();
  });

  it("shows empty state when no notifications", () => {
    const { container } = render(
      <MantineProvider>
        <NotificationCenter notifications={[]} />
      </MantineProvider>,
    );
    expect(container).toMatchSnapshot();
  });
});
