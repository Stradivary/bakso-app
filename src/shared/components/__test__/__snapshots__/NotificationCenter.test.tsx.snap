// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotificationCenter > renders notification center 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="dialog"
    class="m_e5262200 mantine-Indicator-root"
    id=""
    style="--indicator-size: calc(0.75rem * var(--mantine-scale)); --indicator-top: 0rem; --indicator-right: 0rem; --indicator-translate-x: 50%; --indicator-translate-y: -50%;"
  >
    <div
      class="m_760d1fb1 mantine-Indicator-indicator"
      data-processing="true"
    />
    <button
      class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
      data-size="compact-lg"
      data-testid="notification-center"
      data-variant="white"
      style="--button-height: var(--button-height-compact-lg); --button-padding-x: var(--button-padding-x-compact-lg); --button-fz: var(--mantine-font-size-lg); --button-bg: var(--mantine-color-white); --button-hover: rgba(252, 252, 252, 1); --button-color: var(--mantine-color-blue-filled); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
      type="button"
    >
      <span
        class="m_80f1301b mantine-Button-inner"
      >
        <span
          class="m_811560b9 mantine-Button-label"
        >
          <svg
            class="lucide lucide-bell"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 8a6 6 0 0 1 12 0c0 7 3 9 3 9H3s3-2 3-9"
            />
            <path
              d="M10.3 21a1.94 1.94 0 0 0 3.4 0"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`NotificationCenter > shows empty state when no notifications 1`] = `
<div>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 35.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 36em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 61.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 62em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 87.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 88em) {.mantine-hidden-from-xl {display: none !important;}}
  </style>
  <div
    aria-controls=""
    aria-expanded="false"
    aria-haspopup="dialog"
    class="m_e5262200 mantine-Indicator-root"
    id=""
    style="--indicator-size: calc(0.75rem * var(--mantine-scale)); --indicator-top: 0rem; --indicator-right: 0rem; --indicator-translate-x: 50%; --indicator-translate-y: -50%;"
  >
    <button
      class="mantine-focus-auto mantine-active m_77c9d27d mantine-Button-root m_87cf2631 mantine-UnstyledButton-root"
      data-size="compact-lg"
      data-testid="notification-center"
      data-variant="white"
      style="--button-height: var(--button-height-compact-lg); --button-padding-x: var(--button-padding-x-compact-lg); --button-fz: var(--mantine-font-size-lg); --button-bg: var(--mantine-color-white); --button-hover: rgba(252, 252, 252, 1); --button-color: var(--mantine-color-blue-filled); --button-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
      type="button"
    >
      <span
        class="m_80f1301b mantine-Button-inner"
      >
        <span
          class="m_811560b9 mantine-Button-label"
        >
          <svg
            class="lucide lucide-bell"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 8a6 6 0 0 1 12 0c0 7 3 9 3 9H3s3-2 3-9"
            />
            <path
              d="M10.3 21a1.94 1.94 0 0 0 3.4 0"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;
